src/compiler/checker.ts

```diff
@@ -16014,7 +16014,7 @@ namespace ts {
         }
 
         function markAliasReferenced(symbol: Symbol, location: Node) {
-            if (isNonLocalAlias(symbol, /*excludes*/ SymbolFlags.Value) && !isInTypeQuery(location) && !isConstEnumOrConstEnumOnlyModule(resolveAlias(symbol))) {
+            if (isNonLocalAlias(symbol, /*excludes*/ SymbolFlags.Value) && !isInTypeQuery(location) && (compilerOptions.preserveConstEnums || !isConstEnumOrConstEnumOnlyModule(resolveAlias(symbol)))) {

```

Comment:
Don't you need similar change in `markExportAsReferenced` ? Also I think we should remove the `isConstEnumOrConstEnumOnlyModule` check before calling `markAliasSymbolAsReferenced` in `checkJsxOpeningLikeElementOrOpeningFragment`, `checkPropertyAccessExpressionOrQualifiedName`, `markEntityNameOrEntityExpressionAsReference`,

Reply:
> Don't you need similar change in markExportAsReferenced ?

Actually, I don't need to call `markExportAsReferenced` anymore. The `markAliasReferenced` is more correct for an export assignment (as it mimics what `checkIdentifier` would do). `markExportAsReferenced` looks at the wrong symbol for an export assignment (it looks at the `default` export symbol rather than the symbol of the identifier in the expression), so it hasn't been doing any meaningful work anyway. I'll remove its handling of them. (And you know what? The checking functions built into `markAliasSymbolAsReferenced` are strange - I'll try to move those so they're local to what they check, too)

> Also I think we should remove the isConstEnumOrConstEnumOnlyModule check or update it before calling markAliasSymbolAsReferenced in checkJsxOpeningLikeElementOrOpeningFragment,

I don't know why we check const-ness here, actually. I'm pretty sure we just need to always consider it referenced, const or no (since we need to preserve the import because we'll be referencing it for the factory function in the emit).

> checkPropertyAccessExpressionOrQualifiedName

I don't think so - the check there is just checking `prop` to see if it's even possible for `parentSymbol` to be const - a fast path. It calls into `markAliasReferenced` which does the full check.

> markEntityNameOrEntityExpressionAsReference

No need - this is only used for decorator metadata, which I think we'd rather have been the same for preserve const enums/non preserved (and const enums are just going to be mapped to `Number` and their namespaces `Object` anyway - unlike other types which may have necessitated keeping imports for classes of some sort; so we don't want their imports kept in preserve mode, since we won't use them!).